include/runtime-support.h

```diff
@@ -629,6 +631,64 @@ inline ::zeek::ValPtr to_val(const T& t, ::zeek::TypePtr target, const std::stri
     return zeek::compat::ToValPtr(std::move(rval));
 }
 
+/**
+ * Converts Spicy-side struct to a Zeek record value. The result is returned
+ * with a ref count +1.
+ */
+template<typename T, typename std::enable_if_t<std::is_base_of<::hilti::rt::trait::isStruct, T>::value>*>
+inline ::zeek::ValPtr to_val(const T& t, ::zeek::TypePtr target, const std::string& location) {
+    if ( target->Tag() != ::zeek::TYPE_RECORD )
+        throw TypeMismatch("struct", target, location);
+
+    auto rtype = target->AsRecordType();
+
+    auto rval = std::make_unique<::zeek::RecordVal>(zeek::compat::ToValCtorType(rtype));
+    int idx = 0;
+
+    auto numFields = rtype->NumFields();
+
+    t.__visit([&](const auto& name, const auto& val) {
+        if ( idx >= numFields )
+            throw TypeMismatch(hilti::rt::fmt("no matching record field for field '%s'", name), location);
+
+        auto field = zeek::compat::RecordType_GetFieldType(rtype, idx);
+        std::string fieldName = rtype->FieldName(idx);
+
+        if ( fieldName != name )
+            throw TypeMismatch(hilti::rt::fmt("mismatch in field name: expected '%s', found '%s'", name, fieldName),
+                               location);
+
+        ::zeek::ValPtr v = nullptr;
+
+        if constexpr ( std::is_same<decltype(val), const hilti::rt::Null&>::value ) {
+            // "Null" turns into an unset optional record field.
+        }
+        else
+            // This may return a nullptr in cases where the field is to be left unset.
+            v = to_val(val, field, location);
+
+        if ( v )
+            rval->Assign(idx, v);
+        else {
+            // Field must be &optional or &default.
+            auto attrs = rtype->FieldDecl(idx)->attrs;
+            if ( ! (zeek::compat::Attribute_Find(attrs, ::zeek::detail::ATTR_DEFAULT) ||
+                    zeek::compat::Attribute_Find(attrs, ::zeek::detail::ATTR_OPTIONAL)) )
+                throw TypeMismatch(hilti::rt::fmt("missing initialization for field '%s'", fieldName), location);
+        }
+
+        idx++;
+    });
+
+    // We already check above that all Spicy-side fields are mapped so we
+    // can only hit this if there are uninitialized Zeek-side fields left.
+    if ( idx != numFields )
+        throw TypeMismatch(hilti::rt::fmt("missing initialization for field '%s'", rtype->FieldName(idx + 1)),
+                           location);
```

Comment:
I think this good enough as it is, but just to say it, we could be more fancy as well: we could (1) match fields by name so that the struct and record types don't need to have the same ordering, and (2) we could allow fields to be missing if the corresponding record field was optional/defaulted. Implementing that would require more bookkeeping, probably not worth it for now; can always generalize  later.